modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTableLFU.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.siddhi.core.table;
+
+import io.siddhi.core.config.SiddhiAppContext;
+import io.siddhi.core.event.ComplexEventChunk;
+import io.siddhi.core.event.state.StateEvent;
+import io.siddhi.core.event.stream.StreamEvent;
+import io.siddhi.core.table.holder.IndexEventHolder;
+import io.siddhi.core.table.holder.IndexedEventHolder;
+import io.siddhi.core.util.collection.AddingStreamEventExtractor;
+import io.siddhi.core.util.collection.operator.CompiledCondition;
+import io.siddhi.core.util.collection.operator.Operator;
+import io.siddhi.query.api.definition.Attribute;
+import io.siddhi.query.api.definition.TableDefinition;
+
+import static io.siddhi.core.util.SiddhiConstants.CACHE_TABLE_COUNT_LFU;
+import static io.siddhi.core.util.SiddhiConstants.CACHE_TABLE_TIMESTAMP_ADDED;
+import static io.siddhi.core.util.cache.CacheUtils.getPrimaryKey;
+import static io.siddhi.core.util.cache.CacheUtils.getPrimaryKeyFromMatchingEvent;
+
+/**
+ * cache table with FIFO entry removal
+ */
+public class CacheTableLFU extends CacheTable {
+
+    @Override
+    public StreamEvent find(CompiledCondition compiledCondition, StateEvent matchingEvent) {
+        TableState state = stateHolder.getState();
```

Comment:
Better to getState() after the 
readWriteLock.readLock().lock();
as we are returning before releasing.